src/builders/utils/unimodules.ts

```diff
@@ -0,0 +1,162 @@
+    return keys(packageJson.dependencies);
+  }
+  logger.warn('No core packages detected');
+  return [];
+}
+
+class Resolver {
+  private workingdir: string;
+  private modulesMap: {[key: string]: IUnimoduleEntry};
+  private dependencyMap: {[key: string]: IPackageInfo};
+
+  constructor(workingdir: string) {
+    this.workingdir = workingdir;
+    this.modulesMap = {};
+    this.dependencyMap = {};
+  }
+
+  public async init(): Promise<void> {
+    this.dependencyMap = await generateDependencyMap(this.workingdir);
+  }
+
+  public isUnimodule(moduleName: string): boolean {
+    return !!this.dependencyMap[moduleName];
+  }
+
+  public addAll() {
+    for (const moduleName of Object.keys(this.dependencyMap)) {
+      this.addModule(moduleName);
+    }
+  }
+
+  public addModule(moduleName: string) {
+    if (this.dependencyMap[moduleName]) {
+      this.resolveModule(moduleName);
+    } else {
+      throw new Error('Attempting to add unknown module');
+    }
+  }
+
+  public getModules(): IUnimoduleEntry[] {
+     return values(this.modulesMap);
+  }
+
+  private async resolveModule(moduleName: string) {
+    if (this.modulesMap[moduleName]) {
+      return;
+    }
+    const pkg = this.dependencyMap[moduleName];
+    logger.info(`Adding ${moduleName}:${pkg.version}`);
+    this.modulesMap[moduleName] = {
+      name: pkg.name,
+      dirname: pkg.dirname,
+      version: pkg.version,
+    };
+    for (const dep of pkg.dependencies) {
+      this.resolveModule(dep);
+    }
+  }
+}
+
+async function generateDependencyMap(workingdir: string): Promise<{[key: string]: IPackageInfo}> {
+  const dir = await fs.readdir(path.join(workingdir, 'packages'));
+  const map: {[key: string]: IPackageInfo} = {};
+
+  // scrap directory for package metadata
+  for (const pkgName of dir) {
+    const pkg = await getPackage(workingdir, pkgName);
+    if (pkg) {
+      map[pkg.name] = pkg;
+    }
+  }
+
+  // handle @unimodules directory
+  const unimodulesDir = await fs.readdir(path.join(workingdir, 'packages', '@unimodules'));
+  for (const pkgName of unimodulesDir) {
+    const pkg = await getPackage(workingdir, path.join('@unimodules', pkgName));
+    if (pkg) {
+      map[pkg.name] = pkg;
+    }
+  }
+
+  // remove non unimodule dependencies
+  for (const pkg of Object.keys(map)) {
+    map[pkg].dependencies = filter(map[pkg].dependencies, (dep) => !!map[dep]);
+  }
+  return map;
+}
+
+async function getPackage(workingdir: string, pkgName: string): Promise<IPackageInfo | null> {
+  const pkgPath = path.join(workingdir, 'packages', pkgName);
+  const isAndroid = await fs.pathExists(path.join(pkgPath, 'android'));
+  const isIos = await fs.pathExists(path.join(pkgPath, 'ios'));
+  const isUnimodule = await fs.pathExists(path.join(pkgPath, 'unimodule.json'));
+  const hasPackageJson = await fs.pathExists(path.join(pkgPath, 'package.json'));
+  if (!isUnimodule || !hasPackageJson || (!isAndroid && !isIos)) {
```

Comment:
What would happen if one would want to build an app with a unimodule that only supports one platform?

Reply:
enabled-modules dir will contain unsupported dependency, autoinstall script will detect that it's unsupported and it won't be included